apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+
+namespace Google.Cloud.Diagnostics.Common
+{
+    /// <summary>
+    /// An <see cref="IConsumer{T}"/> that will allow for the option to retry RPC failures and buffer
+    /// the data while the retries occur.
+    /// </summary>
+    internal class RpcRetryConsumer<T> : IConsumer<T>
+    {
+        /// <summary>The options dictating how retries will be handled.</summary>
+        private readonly RetryOptions _options;
+
+        /// <summary>A function to obtain the size of an item in bytes.</summary>
+        private readonly Func<T, int> _sizer;
+
+        /// <summary>
+        /// Function to create a <see cref="ISequentialThreadingTimer"/> that will start the given
+        /// action based on retry options.
+        /// </summary>
+        private readonly Func<Action, RetryOptions, ISequentialThreadingTimer> _timerFactory;
+
+        /// <summary>The timer running in the background to retry failed requests.</summary>
+        private readonly ISequentialThreadingTimer _timer;
+
+        /// <summary>A semaphore to protect the buffer.</summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        /// <summary>The consumer to send items to.</summary>
+        internal readonly IConsumer<T> _consumer;
+
+        /// <summary>
+        /// The buffered items. This is not readonly as it is replaced when the buffer is successfully send to
+        /// the base consumer.
+        /// </summary>
+        private List<T> _buffer = new List<T>();
+
+        /// <summary>The size of the buffer.</summary>
+        private int _size = 0;
```

Comment:
Size is ambiguous here (it could be confused with the number of items in the buffer). Perhaps `_sizeInBytes`?